presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -78,30 +82,26 @@ private PickTableLayoutForPredicate(Metadata metadata)
                 return Optional.empty();
             }
 
-            TableLayoutRewriter tableLayoutRewriter = new TableLayoutRewriter(metadata, context.getSession(), context.getSymbolAllocator(), context.getIdAllocator());
-            PlanNode rewrittenTableScan = tableLayoutRewriter.planTableScan((TableScanNode) source, filterNode.getPredicate());
-
-            if (rewrittenTableScan instanceof TableScanNode || rewrittenTableScan instanceof ValuesNode || (((FilterNode) rewrittenTableScan).getPredicate() != filterNode.getPredicate())) {
-                return Optional.of(rewrittenTableScan);
-            }
-
-            return Optional.empty();
+            TableLayoutRewriter tableLayoutRewriter = new TableLayoutRewriter(metadata, context.getSession(), context.getSymbolAllocator(), context.getIdAllocator(), sqlParser);
+            return tableLayoutRewriter.planTableScan((TableScanNode) source, filterNode.getPredicate());
         }
 
         private boolean shouldRewriteTableLayout(TableScanNode source)
```

Comment:
you do need to check `!source.getLayout().isPresent()` because when replaceExistingLayout is false, you still want to add a layout if none existed.  You can't just check `planChanged` then because you only want to return the changed plan if the original plan didn't have a layout.

Reply:
True, but `source.getOriginalConstraint() == BooleanLiteral.TRUE_LITERAL;` can be removed